src/TurbulenceClosures/turbulence_closure_implementations/Smagorinskys/dynamic_coefficient.jl

```diff
@@ -0,0 +1,272 @@
+    Σ̄ = CenterField(grid)
+
+    𝒥ᴸᴹ = Field(Average(LM, dims=closure.coefficient.averaging))
+    𝒥ᴹᴹ = Field(Average(MM, dims=closure.coefficient.averaging))
+
+    return (; Σ, Σ̄, LM, MM, 𝒥ᴸᴹ, 𝒥ᴹᴹ)
+end
+
+#####
+##### Lagrangian-averaged functionality
+#####
+
+const c = Center()
+
+@kernel function _lagrangian_average_LM_MM!(𝒥ᴸᴹ, 𝒥ᴹᴹ, 𝒥ᴸᴹ⁻, 𝒥ᴹᴹ⁻, 𝒥ᴸᴹ_min, Σ, Σ̄, grid, Δt, u, v, w)
+    i, j, k = @index(Global, NTuple)
+    LM, MM = LM_and_MM(i, j, k, grid, Σ, Σ̄, u, v, w)
+    FT = eltype(grid)
+
+    @inbounds begin
+        𝒥ᴸᴹ⁻ᵢⱼₖ = max(𝒥ᴸᴹ⁻[i, j, k], 𝒥ᴸᴹ_min)
+        𝒥ᴹᴹ⁻ᵢⱼₖ = 𝒥ᴹᴹ⁻[i, j, k]
+
+        # Compute time scale
+        𝒥ᴸᴹ𝒥ᴹᴹ = 𝒥ᴸᴹ⁻ᵢⱼₖ * 𝒥ᴹᴹ⁻ᵢⱼₖ
+
+        T⁻ = convert(FT, 1.5) * Δᶠ(i, j, k, grid) / ∜(∜(𝒥ᴸᴹ𝒥ᴹᴹ))
+        τ = Δt / T⁻
+        ϵ = τ / (1 + τ)
+                        
+        # Compute interpolation
+        x = xnode(i, j, k, grid, c, c, c)
+        y = ynode(i, j, k, grid, c, c, c)
+        z = znode(i, j, k, grid, c, c, c)
+
+        # Displacements
+        δx = u[i, j, k] * Δt
+        δy = v[i, j, k] * Δt
+        δz = w[i, j, k] * Δt
+
+        # Prevent displacements from getting too big?
+        Δx = Δxᶜᶜᶜ(i, j, k, grid)
+        Δy = Δyᶜᶜᶜ(i, j, k, grid)
+        Δz = Δzᶜᶜᶜ(i, j, k, grid)
+
+        δx = clamp(δx, -Δx, Δx)
+        δy = clamp(δy, -Δy, Δy)
+        δz = clamp(δz, -Δz, Δz)
+
+        # Previous locations
+        x⁻ = x - δx
+        y⁻ = y - δy
+        z⁻ = z - δz
+        X⁻ = (x⁻, y⁻, z⁻)
+
+        itp_𝒥ᴹᴹ⁻ = interpolate(X⁻, 𝒥ᴹᴹ⁻, (c, c, c), grid)
+        itp_𝒥ᴸᴹ⁻ = interpolate(X⁻, 𝒥ᴸᴹ⁻, (c, c, c), grid)
+
+        # Take time-step
+        𝒥ᴹᴹ[i, j, k] = ϵ * MM + (1 - ϵ) * itp_𝒥ᴹᴹ⁻
+
+        𝒥ᴸᴹ★ = ϵ * LM + (1 - ϵ) * max(itp_𝒥ᴸᴹ⁻, 𝒥ᴸᴹ_min)
+        𝒥ᴸᴹ[i, j, k] = max(𝒥ᴸᴹ★, 𝒥ᴸᴹ_min)
+    end
+end
+
+function compute_coefficient_fields!(diffusivity_fields, closure::LagrangianAveragedDynamicSmagorinsky, model; parameters)
+    grid = model.grid
+    arch = architecture(grid)
+    clock = model.clock
+    cˢ = closure.coefficient
+    t⁻ = diffusivity_fields.previous_compute_time
+    u, v, w = model.velocities
+
+    Δt = clock.time - t⁻[]
+    t⁻[] = model.clock.time
+
+    if cˢ.schedule(model)
+        Σ = diffusivity_fields.Σ
+        Σ̄ = diffusivity_fields.Σ̄
+        launch!(arch, grid, :xyz, _compute_Σ_Σ̄!, Σ, Σ̄, grid, u, v, w)
+
+        parent(diffusivity_fields.𝒥ᴸᴹ⁻) .= parent(diffusivity_fields.𝒥ᴸᴹ)
+        parent(diffusivity_fields.𝒥ᴹᴹ⁻) .= parent(diffusivity_fields.𝒥ᴹᴹ)
+
+        𝒥ᴸᴹ⁻ = diffusivity_fields.𝒥ᴸᴹ⁻
+        𝒥ᴹᴹ⁻ = diffusivity_fields.𝒥ᴹᴹ⁻
+        𝒥ᴸᴹ  = diffusivity_fields.𝒥ᴸᴹ
+        𝒥ᴹᴹ  = diffusivity_fields.𝒥ᴹᴹ
+        𝒥ᴸᴹ_min = cˢ.minimum_numerator
+
+        if !isfinite(clock.last_Δt) || Δt == 0 # first time-step
+            @info "               Launching _compute_LM_MM! at t=0"
+            launch!(arch, grid, :xyz, _compute_LM_MM!, 𝒥ᴸᴹ, 𝒥ᴹᴹ, Σ, Σ̄, grid, u, v, w)
+            @info "               Finished _compute_LM_MM!"
```

Comment:
The buildkite tests are getting stuck here, compiling `_compute_LM_MM!()`. I found that if I comment out some stuff either in `_compute_LM_MM!()` or in `LM_and_MM()` (the former calls the latter) then they compile and tests pass.

I have tried this with different local machines, and tests pass in all of them. So I'm really not sure what's different about the buildkite set-up that's causing this.

Reply:
on GPU or CPU?

Reply:
eg on tartarus (CPU) or sverdrup (GPU)?

If one of them isn't up to the task, let's just mark the test as `@test_broken`

Reply:
I've been focusing on CPU tests, but I can also investigate this on the GPU.

What do you mean "up to the task" though? I thought for CPU these things should compile regardless of the machine or kernel complexity, no?

Reply:
Compilation still has a computational cost!

Reply:
Hmm, it seems neither the CPU nor GPU turbulence tests compile `_compute_LM_MM!()`. I just re-started the time-stepping tests so that we see if it compiles there. I'm reluctant to merge this PR without any sort of working tests for the dynamic closures.

Reply:
This PR does much more than just implement the dynamic closures. In terms of going out of date, the outcome is the same whether we merge the PR or not. It's definitely better to merge than to let it sit around growing stale.

Reply:
Focus on productivity, development, and moving forward.

Reply:
> Add a hook that avoids the computation of LM_MM, and use it in the test. Then there is some kind of unit test which is only missing the LM_MM computation (which is very unlikely to break). Then we can use the closure.

Ah, I missed this suggestion! I'll open another PR later to implement this.

Can you clarify what you mean by hook though? Do you mean an extra flag for the dynamic coefficient that skips that computation?